### PR TITLE
Task3 new

### PR DESCRIPTION
**Pull request by ildarzmtv**
What has been done in this work:
- EDA by attribute groups
- Scaling method selection (StandardScaler was used)
- Feature selection methods comparison (RFE with 150 features was used)

The final model scores:
- Accuracy: 0.862745 
- Recall: 0.964912
- Precision: 0.866142
- F1: 0.912863

Confusion matrix: 

image

### CODE REVIEW

**Review by egorponomarev — COMMENTED** [review]
.py и .ipynb явно не в синке. В ipynb нескольок рантайм ошибок:
image
image

Общий флоу не особо понимаю, за обучением следует feature selection, scaling и ещё одно обучение. Буду считать что что-то не то отправлено

File: Task3_new.py [hunk 1]
```diff
@@ -0,0 +1,779 @@
...
+# For non-robust methods i will try: StandardScaler, MinMaxScaler  
+# For robust methods: QuantileTransformer, RobustScaler
+# 
+
+# Some helper functions:
+
+# In[27]:
+
+
+def get_scores(name, y_train, y_pred_train, y_test, y_pred_test):
```
Comment (egorponomarev) [inline Task3_new.py:493]: Нам вчера показали тулзу [mlflow](https://mlflow.org/), по идее она делает то, что у тебя тут написано

File: Task3_new.py [hunk 2]
```diff
@@ -0,0 +1,779 @@
...
+pipeline = Pipeline([
+    ('scaler', StandardScaler()),
+    ('clf', LogisticRegression(random_state=RANDOM_STATE))
+])
+
+
+# In[29]:
+
+
+weights_for_0_class = np.linspace(0.5, 0.6, 20)
```
Comment (egorponomarev) [inline Task3_new.py:601]: Выглядит не оптимально. Дисбаланс классов в 4 раза, значит логично подбирать в промежутке, включающем 0.8 (или 0.75...). А если передать в логистическую регрессию `class_weight: 'balanced'` то автоматически подберутся веса, соответствующие распределению таргета

File: Task3_new.py [hunk 3]
```diff
@@ -0,0 +1,779 @@
...
+
+# Our dataset consists of 755 features, and there are definetly redundant ones. Some of them are correlated with each other, others are just not good predictors.  
+# 
+# To reduce the dimensionality, I'll try 2 feature selection methods - Recursive Feature Elimination (RFE) and Maximum Relevance — Minimum Redundancy (MRMR).  
+# RFE is popular because it is easy to configure and use and because it is effective at selecting the most relevantfeatures  in predicting the target variable.  
+# The MRMR approach is based on maximizing the joint dependency of top ranking variables on the target variable by reducing the redundancy among them. [MRMR explanation]("https://towardsdatascience.com/mrmr-explained-exactly-how-you-wished-someone-explained-to-you-9cf4ed27458b") 
+
+# For each feature selection method i tuned models with 50, 100, 150, 200, 250 features and that is the results:
+
+# ![image.png](attachment:bde9d848-99bb-401b-8d45-f8e6541d02f6.png)
```
Comment (egorponomarev) [inline Task3_new.py:663]: Что за прикол с картинками... Как мы должны их валидировать без кода

File: Task3_new.py [hunk 4]
```diff
@@ -0,0 +1,779 @@
...
+    X_test : Test set
+    y_test : Test set labels
+    
+    Returns
+    -------
+    Function returns the dataframe with accuracy, recall, precision and f1 scores 
+    for train and test sets
+    """
+    
+    model.fit(X_train, y_train)
```
Comment (egorponomarev) [inline Task3_new.py:550]: Вроде не возникает проблемы в коде (возможно в зоне с GridSearch), но опять fit на модели, которая потенциально уже обучена

File: Task3_new.py (hunk 2 above)
Comment (ildarzmtv) [reply Task3_new.py:601]: При таком балансе падает Recall первого класса, поэтому веса именно таким образом подобраны. Хороший recall, и всё ещё не плохой Precision.

File: Task3_new.py (hunk 3 above)
Comment (ildarzmtv) [reply Task3_new.py:663]: Подумал, что большой кусок с тюнингом разных вариантов моделей не особо интересен. В качестве примера там есть один флоу, через который собственно я и прогнал остальные модели. Сеттинг модели -> Тюнинг -> Добавить в сравнение. Параметры в тюнинге уже подбирал отдельно для каждой модели, пока не сойдутся.
Как в таком случае лучше поступить? Оставлять это?

File: Task3_new.py (hunk 4 above)
Comment (ildarzmtv) [reply Task3_new.py:550]: В sklearn повторный вызов `fit()` начинает обучение заново. [Источник](https://stackoverflow.com/questions/49841324/what-does-calling-fit-multiple-times-on-the-same-model-do)

File: Task3_new.py (hunk 2 above)
Comment (egorponomarev) [reply Task3_new.py:601]: Ну значит он бы нашёлся в GridSearch :) Я не спорю что может быть не оптимально, я говорю что не очевидно

File: Task3_new.py (hunk 3 above)
Comment (egorponomarev) [reply Task3_new.py:663]: Я б убрал, не добавляет к повествовательной части ничего

File: Task3_new.py (hunk 4 above)
Comment (egorponomarev) [reply Task3_new.py:550]: Да, я почему-то всегда думал наоборот 😅

**Review by egorponomarev — APPROVED** [review]
